src/protocol/sort/bit_permutations.rs

```diff
@@ -0,0 +1,156 @@
+    /// Create an object to generate bit permutations for a given bit column of query. This is GENBITPERM(Algorithm 3) from the paper
+    #[allow(dead_code)]
+    pub fn new(input: &'a [Replicated<F>]) -> BitPermutations<'a, F> {
+        Self { input }
+    }
+
+    /// In this step, multiplication inputs are generated locally at each helper from input share, x in following steps
+    /// 1. calculate 1 - x, x and concatenate them
+    /// 2. calculate cumulative sum at each vector row
+    /// 3. return back tuple of step 1 and step 2 output
+    fn prepare_mult_inputs<M: Mesh, G: Gateway<M, IPAProtocolStep>>(
+        &self,
+        ctx: &ProtocolContext<'_, G, IPAProtocolStep>,
+    ) -> Vec<(Replicated<F>, Replicated<F>)>
+    where
+        F: Field,
+    {
+        let channel = ctx
+            .gateway
+            .get_channel(IPAProtocolStep::Sort(SortStep::BitPermutations));
+
+        self.input
+            .iter()
+            .map(|x| channel.share_of_one() - *x)
+            .chain(self.input.iter().copied())
+            .scan(Replicated::<F>::new(F::ZERO, F::ZERO), |sum, n| {
+                *sum = *sum + n;
+                Some((n, *sum))
+            })
+            .collect()
+    }
+
+    /// multiplies the input vector pairs across helpers and returns result
+    /// For this, it spawns all multiplication, wait for them to finish in parallel and then collect the results
+
+    #[allow(clippy::cast_possible_truncation)]
+    async fn secure_multiply<M: Mesh, G: Gateway<M, IPAProtocolStep>>(
+        &self,
+        ctx: &ProtocolContext<'_, G, IPAProtocolStep>,
+        mult_inputs: Vec<(Replicated<F>, Replicated<F>)>,
+    ) -> Result<Vec<Replicated<F>>, BoxError>
+    where
+        F: Field,
+    {
+        let async_multiply = mult_inputs
+            .iter()
+            .enumerate()
+            .map(|(index, input)| async move {
+                ctx.multiply(
+                    RecordId::from(index as u32),
+                    IPAProtocolStep::Sort(SortStep::BitPermutations),
+                )
+                .await
+                .execute(input.0, input.1)
+                .await
+            });
+
+        try_join_all(async_multiply).await
+    }
+
+    /// Executes sorting of a bit column on mpc helpers. Each helper receives their input shares and do following steps
+    /// 1. local computation by `prepare_mult_inputs` which outputs 2 vectors [x,y]
+    /// 2. multiply each row of previous output individually (i.e. x*y) across mpc helpers.
+    /// 3. add ith column by i+len to obtain helper's share of sorted location, where len is same as input shares length
+    #[allow(dead_code)]
+    pub async fn execute<M: Mesh, G: Gateway<M, IPAProtocolStep>>(
+        self,
+        ctx: &ProtocolContext<'_, G, IPAProtocolStep>,
+    ) -> Result<Vec<Replicated<F>>, BoxError>
+    where
+        F: Field,
+    {
+        let mult_input = self.prepare_mult_inputs(ctx);
```

Comment:
I stared at this code again and I believe there is a way to avoid buffering. What happens here is `mult_input` is a vector of size `2*N` and `mult_output` is a vector as well (same size). so in line 93 we have 4*N items in memory. we can do better by returning an iterator from `prepare_mult_inputs` like this


```rust
 fn prepare_mult_inputs<M: Mesh, G: Gateway<M, IPAProtocolStep>>(
        &self,
        ctx: &ProtocolContext<'_, G, IPAProtocolStep>,
    ) -> impl Iterator<Item=(Replicated<F>, Replicated<F>)> + 'a
    where
        F: Field,
    {
        let share_of_one = ctx.gateway
            .get_channel(IPAProtocolStep::Sort(SortStep::BitPermutations))
            .share_of_one();

        self.input
            .iter()
            .map(move |x| share_of_one - *x)
            .chain(self.input.iter().copied())
            .scan(Replicated::<F>::new(F::ZERO, F::ZERO), |sum, n| {
                *sum = *sum + n;
                Some((n, *sum))
            })
    }
```

`secure_multiply` just need to take an iterator instead of `Vec`

```rust
async fn secure_multiply<M: Mesh, G: Gateway<M, IPAProtocolStep>>(
        &self,
        ctx: &ProtocolContext<'_, G, IPAProtocolStep>,
        mult_inputs: impl Iterator<Item=(Replicated<F>, Replicated<F>)>,
    ) -> Result<Vec<Replicated<F>>, BoxError>
```

Reply:
Have made the change such that at no given time we have 4N memory. 
prepare_mult_inputs returns an iterator. We run a map on top of it to call secure_multiply for each value (which will also return in an iterator). This returns an async handle which we collect in a vector. In the end, join_all is run on all async handles on multiplications. LMK if this looks fine